src/par_iter/mod.rs

```diff
@@ -468,35 +469,70 @@ pub trait ParallelIterator: Sized {
         reduce(self, PRODUCT)
     }
 
-    /// Computes the minimum of all the items in the iterator.
+    /// Computes the minimum of all the items in the iterator. If the
+    /// iterator is empty, `None` is returned; otherwise, `Some(min)`
+    /// is returned.
     ///
-    /// Note that the order in items will be reduced is not specified,
-    /// so if the `Ord` impl is not truly associative, then the
-    /// results are not deterministic.
+    /// Note that the order in which the items will be reduced is not
+    /// specified, so if the `Ord` impl is not truly associative, then
+    /// the results are not deterministic.
     ///
-    /// Basically equivalent to `self.reduce(|| MAX, |a, b| cmp::min(a, b))`
-    /// except that the type of `MAX` and the `min` operation may vary
-    /// depending on the type of value being produced.
-    fn min(self) -> Self::Item
-        where MinOp: ReduceOp<Self::Item>
+    /// Basically equivalent to `self.reduce_with(|a, b| cmp::min(a, b))`.
+    fn min(self) -> Option<Self::Item>
+        where Self::Item: Ord
     {
-        reduce(self, MIN)
+        self.reduce_with(cmp::min)
     }
 
-    /// Computes the maximum of all the items in the iterator.
+    /// Computes the item that yields the minimum value for the given
+    /// function. If the iterator is empty, `None` is returned;
+    /// otherwise, `Some(item)` is returned.
     ///
-    /// Note that the order in items will be reduced is not specified,
-    /// so if the `Ord` impl is not truly associative, then the
-    /// results are not deterministic.
+    /// Note that the order in which the items will be reduced is not
+    /// specified, so if the `Ord` impl is not truly associative, then
+    /// the results are not deterministic.
+    fn min_by_key<K, F>(self, f: F) -> Option<Self::Item>
+        where K: Ord,
+              F: Sync + Fn(&Self::Item) -> K,
+    {
+        self.reduce_with(|a, b| match f(&a).cmp(&f(&b)) {
+            Ordering::Less => a,
+            Ordering::Equal=> b,
```

Comment:
I *think* we want `a` in the case where they are equal, to preserve stability. At least this is what I think sequential iterators do. Can you edit `src/par_iter/test.rs` and add some tests for `min_by_key`, and specifically check for cases where two things compare equal?

Something like this, but probably with a bigger array :)

```rust
#[test]
fn stable_min_by_key() {
    let x = vec![(0, 'a'), (22, 'b'), (0, 'c')];
    assert_eq!(
        x.par_iter().weight_max().min_by_key(|pair| pair.0),
        x.iter().min_by_key(|pair| pair.0)
    );
}
```